ldm/modules/kohya_lora_manager.py

```diff
@@ -0,0 +1,375 @@
+                layer.mid = layer_mid
+                layer.up = layer_up
+
+            # loha
+            elif "hada_w1_b" in values:
+
+                rank = values["hada_w1_b"].shape[0]
+
+                layer = LoHALayer(self.name, stem, rank, alpha)
+                layer.org_module = wrapped
+                layer.bias = bias
+
+                layer.w1_a = values["hada_w1_a"].to(device=self.device, dtype=self.dtype)
+                layer.w1_b = values["hada_w1_b"].to(device=self.device, dtype=self.dtype)
+                layer.w2_a = values["hada_w2_a"].to(device=self.device, dtype=self.dtype)
+                layer.w2_b = values["hada_w2_b"].to(device=self.device, dtype=self.dtype)
+
+                if "hada_t1" in values:
+                    layer.t1 = values["hada_t1"].to(device=self.device, dtype=self.dtype)
+                else:
+                    layer.t1 = None
+
+                if "hada_t2" in values:
+                    layer.t2 = values["hada_t2"].to(device=self.device, dtype=self.dtype)
+                else:
+                    layer.t2 = None
+
+            else:
+                print(
+                    f">> Encountered unknown lora layer module in {self.name}: {stem} - {type(wrapped).__name__}"
+                )
+                return
+
+            self.layers[stem] = layer
+
+
+class KohyaLoraManager:
+    def __init__(self, pipe, lora_path):
+        self.unet = pipe.unet
+        self.lora_path = lora_path
+        self.wrapper = LoRAModuleWrapper(pipe.unet, pipe.text_encoder)
+        self.text_encoder = pipe.text_encoder
+        self.device = torch.device(choose_torch_device())
+        self.dtype = pipe.unet.dtype
+        self.loras_to_load = {}
+
+    def load_lora_module(self, name, path_file, multiplier: float = 1.0):
+        print(f"   | Found lora {name} at {path_file}")
+        if path_file.suffix == ".safetensors":
+            checkpoint = load_file(path_file.absolute().as_posix(), device="cpu")
+        else:
+            checkpoint = torch.load(path_file, map_location="cpu")
+
+        lora = LoRA(name, self.device, self.dtype, self.wrapper, multiplier)
+        lora.load_from_dict(checkpoint)
+        self.wrapper.loaded_loras[name] = lora
+
+        return lora
+
+    def apply_lora_model(self, name, mult: float = 1.0):
+        for suffix in ["ckpt", "safetensors", "pt"]:
+            path_file = Path(self.lora_path, f"{name}.{suffix}")
+            if path_file.exists():
+                print(f"   | Loading lora {path_file.name} with weight {mult}")
+                break
+        if not path_file.exists():
+            print(f"   ** Unable to find lora: {name}")
+            return
+
+        lora = self.wrapper.loaded_loras.get(name, None)
+        if lora is None:
+            lora = self.load_lora_module(name, path_file, mult)
+
+        lora.multiplier = mult
+        self.wrapper.applied_loras[name] = lora
+
+    def unload_applied_loras(self, loras_to_load):
```

Comment:
This function is unused.
Also for me such definition controversial to `unload_applied_lora(lora_name)` definition.

Reply:
Correct. We don't call this. I expect it can be removed as it is probably dangerous to call.

Reply:
Both issues addressed in recent commit.